### ISSUE TITLE
Can't handle Symbol in query with CASE and regexp_matches

### ISSUE THREAD

**Issue by hlcianfagna** [crate/crate#16868]
### CrateDB version

5.8.3 and `nightly-5.10.0-2024-10-24-00-02`

### CrateDB setup information

Reproed both on single node and with 3 nodes

### Problem description

The query below fails if `field2` which is involved in the `CASE` is not `SELECT`ed

### Steps to Reproduce

```sql
 CREATE TABLE test (	data OBJECT );
 
 INSERT INTO test VALUES
 (
 $$
	{				
		"field1":"abc",
		"field2":"def"
	}
 $$
 );
 
REFRESH TABLE test;
 
SELECT data ['field1'],
	data['field2'] AS field2,
	CASE 
		WHEN regexp_matches(data ['field1'], '^a') IS NOT NULL
			THEN 'found'
		WHEN data ['field2'] LIKE '%xyz%'
			THEN 'special case'
		ELSE 'default'
		END
FROM test;
--> works

SELECT data ['field1'],
	--data['field2'] AS field2,
	CASE 
		WHEN regexp_matches(data ['field1'], '^a') IS NOT NULL
			THEN 'found'
		WHEN data ['field2'] LIKE '%xyz%'
			THEN 'special case'
		ELSE 'default'
		END
FROM test;
```

### Actual Result

```
io.crate.exceptions.UnsupportedFeatureException: Can't handle Symbol [SimpleReference: data['field2']]
	at io.crate.exceptions.SQLExceptions.esToCrateException(SQLExceptions.java:213)
	at io.crate.exceptions.SQLExceptions.prepareForClientTransmission(SQLExceptions.java:200)
	at io.crate.rest.action.SqlHttpHandler.sendResponse(SqlHttpHandler.java:172)
	at io.crate.rest.action.SqlHttpHandler.lambda$channelRead0$0(SqlHttpHandler.java:127)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:931)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2401)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:125)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:86)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.protocols.http.HttpBlobHandler.channelRead0(HttpBlobHandler.java:159)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.auth.HttpAuthUpstreamHandler.handleHttpChunk(HttpAuthUpstreamHandler.java:155)
	at io.crate.auth.HttpAuthUpstreamHandler.channelRead0(HttpAuthUpstreamHandler.java:92)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:289)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at org.elasticsearch.transport.netty4.Netty4InboundStatsHandler.channelRead(Netty4InboundStatsHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: java.lang.UnsupportedOperationException: Can't handle Symbol [SimpleReference: data['field2']]
	at io.crate.expression.BaseImplementationSymbolVisitor.visitSymbol(BaseImplementationSymbolVisitor.java:103)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitSymbol(BaseImplementationSymbolVisitor.java:43)
	at io.crate.expression.symbol.SymbolVisitor.visitReference(SymbolVisitor.java:38)
	at io.crate.metadata.SimpleReference.accept(SimpleReference.java:234)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitFunction(BaseImplementationSymbolVisitor.java:65)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitFunction(BaseImplementationSymbolVisitor.java:43)
	at io.crate.expression.symbol.Function.accept(Function.java:249)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitFunction(BaseImplementationSymbolVisitor.java:65)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitFunction(BaseImplementationSymbolVisitor.java:43)
	at io.crate.expression.symbol.Function.accept(Function.java:249)
	at io.crate.expression.InputFactory$Context.add(InputFactory.java:159)
	at io.crate.expression.InputFactory$Context.add(InputFactory.java:144)
	at io.crate.expression.InputFactory.ctxForInputColumns(InputFactory.java:98)
	at io.crate.execution.engine.pipeline.ProjectionToProjectorVisitor.visitEvalProjection(ProjectionToProjectorVisitor.java:322)
	at io.crate.execution.engine.pipeline.ProjectionToProjectorVisitor.visitEvalProjection(ProjectionToProjectorVisitor.java:144)
	at io.crate.execution.dsl.projection.EvalProjection.accept(EvalProjection.java:103)
	at io.crate.execution.dsl.projection.ProjectionVisitor.process(ProjectionVisitor.java:30)
	at io.crate.execution.engine.pipeline.ProjectionToProjectorVisitor.create(ProjectionToProjectorVisitor.java:746)
	at io.crate.execution.engine.pipeline.Projectors.<init>(Projectors.java:59)
	at io.crate.execution.engine.collect.sources.ShardCollectSource.getIterator(ShardCollectSource.java:248)
	at io.crate.execution.engine.collect.MapSideDataCollectOperation.createIterator(MapSideDataCollectOperation.java:58)
	at io.crate.execution.engine.collect.CollectTask.start(CollectTask.java:161)
	at io.crate.execution.jobs.RootTask.start(RootTask.java:203)
	at io.crate.execution.jobs.RootTask.start(RootTask.java:188)
	at io.crate.execution.engine.JobLauncher.setupTasks(JobLauncher.java:270)
	at io.crate.execution.engine.JobLauncher.execute(JobLauncher.java:174)
	at io.crate.planner.operators.LogicalPlanner.executeNodeOpTree(LogicalPlanner.java:667)
	at io.crate.planner.operators.LogicalPlanner.doExecute(LogicalPlanner.java:605)
	at io.crate.planner.operators.LogicalPlanner.execute(LogicalPlanner.java:570)
	at io.crate.planner.operators.LogicalPlan.executeOrFail(LogicalPlan.java:216)
	at io.crate.planner.Plan.execute(Plan.java:73)
	at io.crate.session.Session.singleExec(Session.java:797)
	at io.crate.session.Session.execute(Session.java:561)
	at io.crate.rest.action.SqlHttpHandler.executeSimpleRequest(SqlHttpHandler.java:269)
	at io.crate.rest.action.SqlHttpHandler.handleSQLRequest(SqlHttpHandler.java:216)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:124)
	... 50 more
```

### Expected Result

```
+----------------+--------+-------------------------------------------------------------------------------------------------------------------------------------+
| data['field1'] | case(true, 'default', (NOT (regexp_matches(data['field1'], '^a') IS NULL)), 'found', (data['field2'] LIKE '%xyz%'), 'special case') |
+----------------+--------+-------------------------------------------------------------------------------------------------------------------------------------+
| abc            | found
                                    |
+----------------+--------+-------------------------------------------------------------------------------------------------------------------------------------+
```

**Comment by BaurzhanSakhariev** [comment]
Simpler reproduction
```
create table t(x int);
insert into t(x) values (1);
```

```
select 
 case 
  when x = 0 then 'zero'
  when x % 2 = 0 then 'even'
  else 'odd' 
 end
 from t;
```

To make it work we have to add a column inside CASE to the select list.
Similar to how adding `data['field2'] AS field2` fixes the query.

In postgres it works https://www.db-fiddle.com/f/mc5W3oUw71acMKRY5yEAD/0

Even simpler reproduction
`select abs(x) from t`

**Comment by BaurzhanSakhariev** [comment]
On a separate note: regexp_matches is a table function and in case of no matches it returns an empty array, see https://cratedb.com/docs/crate/reference/en/latest/general/builtins/table-functions.html#examples

`regexp_matches(data ['field1'], '^a') IS NOT NULL` must be replaced to `regexp_matches(data ['field1'], '^a') != []`.

BTW, usage of the table function makes LogicalPlanner to use ProjectSet, which ignores `data ['field2']` used in CASE and also leads to using Eval with source = ProjectSet not having `data ['field2']` in outputs.

**Comment by BaurzhanSakhariev** [comment]
Thanks for reporting @hlcianfagna, fix will be available in the 5.9.3 hotfix release.

Thanks @seut for suggesting a better approach!
